File: Data Structures & Algorithms/maximum-subarray/submission-1.py
class Solution:
    def maxSubArray(self, nums: List[int]) -> int:
        maxSum = -float('inf')
        summ = 0

        for n in nums:
            if summ < 0:
                summ = 0
            summ += n
            maxSum = max(maxSum, summ)

        return maxSum

File: Data Structures & Algorithms/maximum-product-subarray/submission-1.cpp
class Solution {
public:
    int maxProduct(vector<int>& nums) {
        int ret = nums[0];
        int curMin = 1, curMax = 1;

        for (int n : nums) {
            if (n == 0) {
                curMin = 1; curMax = 1;
                ret = max(ret, 0);
                continue;
            }

            int temp = curMin * n;
            curMin = min({n * curMin, n * curMax, n});
            curMax = max({temp, n * curMax, n});

            ret = max(ret, curMax);
        }

        return ret;
    }
};
